eventhubs/receive_eph.go

```diff
@@ -0,0 +1,114 @@
+	storage "github.com/Azure-Samples/azure-sdk-for-go-samples/storage"
+)
+
+func ReceiveViaEPH() {
+	ctx := context.Background()
+
+	// get a struct with metadata about the Azure environment
+	azureEnv, err := azure.EnvironmentFromName("AzurePublicCloud")
+	if err != nil {
+		log.Fatalf("could not get azure.Environment struct: %s\n", err)
+	}
+
+	// create an access token provider using AAD principal defined in environment
+	tokenProvider, err := aad.NewJWTProvider(aad.JWTProviderWithEnvironmentVars())
+	if err != nil {
+		log.Fatalf("failed to configure AAD JWT provider: %s\n", err)
+	}
+
+	// create a storage account and container to maintain dictionary of leases
+	// and checkpoints
+	log.Printf("creating storage account\n")
+	_, err = storage.CreateStorageAccount(ctx, storageAccountName)
+	if err != nil {
+		log.Fatalf("could not create storage account: %s\n", err)
+	}
+	log.Printf("creating storage container\n")
+	_, err = storage.CreateContainer(ctx, storageAccountName, storageContainerName)
+	if err != nil {
+		log.Fatalf("could not create storage container: %s\n", err)
+	}
+
+	// use helper method to exchange AAD credentials for SAS token
+	cred, err := storageLeaser.NewAADSASCredential(
+		helpers.SubscriptionID(),
+		helpers.ResourceGroupName(),
+		storageAccountName,
+		storageContainerName,
+		storageLeaser.AADSASCredentialWithEnvironmentVars())
+	if err != nil {
+		log.Fatalf("could not prepare a storage credential: %s\n", err)
+	}
+
+	// create a leaser and checkpointer backed by a storage container
+	leaserCheckpointer, err := storageLeaser.NewStorageLeaserCheckpointer(
+		cred,
+		storageAccountName,
+		storageContainerName,
+		azureEnv)
+	if err != nil {
+		log.Fatalf("could not prepare a storage leaserCheckpointer: %s\n", err)
+	}
+
+	// use all of the above to create an Event Processor
+	p, err := eph.New(
+		ctx,
+		nsName,
+		hubName,
+		tokenProvider,
+		leaserCheckpointer,
+		leaserCheckpointer,
+		eph.WithNoBanner())
+	if err != nil {
+		log.Fatalf("failed to create EPH: %s\n", err)
+	}
+	defer p.Close(context.Background())
+
+	// set up a handler for the Event Processor which will receive a single
+	// message, print it to the console, and allow the process to exit
+
+	// set up wait group to wait for expected message
+	var wg sync.WaitGroup
+	wg.Add(1) // wait for one message
+
+	handler := func(ctx context.Context, event *eventhubs.Event) error {
+		fmt.Printf("received: %s\n", string(event.Data))
+		wg.Done()
+		return nil
+	}
+
+	// register the handler with the Event Processor
+	close, err := p.Receive(handler)
+	defer close()
+	if err != nil {
+		log.Fatalf("failed to set up handler: %s\n", err)
+	}
+
+	// finally, start the Event Processor with a timeout
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	err = p.StartNonBlocking(ctx)
+	if err != nil {
+		log.Fatalf("failed to start EPH: %s\n", err)
+	}
+
+	// don't exit till a message has been processed
+	wg.Wait()
```

Comment:
One of the benefits of switch to using channels instead of wait group would mean you could do this:

``` Go
switch {
case <-receivedEvent:
    // Intentionally Left Blank
case err := <-ctx.Done():
    log.Fatal("Event not received quickly enough:", err )
}
```